### PR TITLE
fix: restore SelfKeepAlive in Menu

### LINKED ISSUE #52247: Replacing menu while open causes segfault on Windows
### Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

43.0.0

### What operating system(s) are you using?

Windows

### Operating System Version

Windows 10.0.19045 (6466)

### What arch are you using?

x64

### Last Known Working Electron version

41.2.0

### Does the issue also appear in Chromium / Google Chrome?

No

### Expected Behavior

Calling `Menu.setApplicationMenu()` while the user interacts with an open menu should not crash the app.

### Actual Behavior

If the application menu is replaced while being displayed, and a GC triggers before the menu closes, the relevant `Menu` object will be garbage-collected, causing various kinds of segfaults in menu-related UI code.

For example:
```
OS Version: Windows 10.0.19045 (6466)
Report Version: 104

Crashed Thread: 38224

Application Specific Information:
Fatal Error: EXCEPTION_ACCESS_VIOLATION_READ / 0x0

Thread 38224 Crashed:
0   Figma.exe                       0x7ff6274073c7      ui::MenuModel::GetModelAndIndexForCommandId (menu_model.cc:49)
1   Figma.exe                       0x7ff62a4fd728      views::MenuModelAdapter::GetLabelFontList (menu_model_adapter.cc:225)
2   Figma.exe                       0x7ff62a507cdf      views::MenuItemView::GetFontList (menu_item_view.cc:1005)
3   Figma.exe                       0x7ff62a505f35      views::MenuItemView::OnPaintImpl (menu_item_view.cc:1115)
4   Figma.exe                       0x7ff620987cfb      views::View::Paint (view.cc:1475)
5   Figma.exe                       0x7ff620988261      views::View::RecursivePaintHelper (view.cc:3114)
6   Figma.exe                       0x7ff62098811f      views::View::PaintChildren (view.cc:2436)
7   Figma.exe                       0x7ff62a502328      views::SubmenuView::PaintChildren (submenu_view.cc:316)
8   Figma.exe                       0x7ff620987d6f      views::View::Paint (view.cc:1481)
9   Figma.exe                       0x7ff620988261      views::View::RecursivePaintHelper (view.cc:3114)
10  Figma.exe                       0x7ff62098811f      views::View::PaintChildren (view.cc:2436)
11  Figma.exe                       0x7ff620987d6f      views::View::Paint (view.cc:1481)
12  Figma.exe                       0x7ff620988261      views::View::RecursivePaintHelper (view.cc:3114)
13  Figma.exe                       0x7ff62098811f      views::View::PaintChildren (view.cc:2436)
14  Figma.exe                       0x7ff620987d6f      views::View::Paint (view.cc:1481)
15  Figma.exe                       0x7ff620988261      views::View::RecursivePaintHelper (view.cc:3114)
16  Figma.exe                       0x7ff62098811f      views::View::PaintChildren (view.cc:2436)
17  Figma.exe                       0x7ff620987d6f      views::View::Paint (view.cc:1481)
18  Figma.exe                       0x7ff620988261      views::View::RecursivePaintHelper (view.cc:3114)
19  Figma.exe                       0x7ff62098811f      views::View::PaintChildren (view.cc:2436)
20  Figma.exe                       0x7ff620987d6f      views::View::Paint (view.cc:1481)
21  Figma.exe                       0x7ff622ce48fe      views::View::PaintFromPaintRoot (view.cc:3122)
22  Figma.exe                       0x7ff622d60191      ui::Layer::PaintContentsToDisplayList (layer.cc:1599)
```

### Testcase Gist URL

https://gist.github.com/rf-figma/103e9c2d491975cbefe183bfa5ca3cc5

### Additional Information

To trigger the issue in Fiddle, you need to run the app with `--js-flags=--expose-gc`, open the File menu, and mouse over the menu a bit.

This issue was introduced in 41.2.1, almost definitely in #50915 (the backport of #50465).

That PR contains a bunch of GC-related changes, one of which was the removal of a `gin_helper::SelfKeepAlive` property from `Menu`. I think [the premise of the change was that this was redundant](https://github.com/electron/electron/pull/50806#pullrequestreview-4079112260), since `Menu` instances are all owned by other objects that should retain them properly.

The problem with this is that the owner of a `Menu` will only retain it as long as it hasn't been replaced. If you replace the menu while it's open, and the `Menu` object gets garbage-collected as a result, its destructor will clear all of the delegate pointers that are still being used by the rendered views.

I tried just restoring the `SelfKeepAlive` and it did fix the issue for me, although I'm not sure if there's some reason that we would need to do this differently now. I'll put up a PR in a second.

### PR DESCRIPTION
**Pull request by rf-figma**
#### Description of Change

This fixes #52247 by ensuring that a `Menu` instance will stay alive as long as the menu is open, even if its owner has cleared or replaced it.

It's a (very) partial revert of 9be03cbe54697, which removed this `keep_alive_` property because it was thought to be unnecessary.

Since 9be03cbe54697 didn't need to change any tests I'm assuming there are none for this aspect of the system, but I'd definitely be open to suggestions about where they could be added.

#### Checklist

- [x] I have built and tested this change
- [x] I have filled out the PR description
- [x] `npm test` passes — there were a handful of unrelated-looking failures on my machine
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Fixed crash triggered by replacing an open application menu.

### CODE REVIEW

**Comment by rf-figma** [comment]
Note that I tested this on Fiddle against `v42.5.0`, but I haven't tested it against `main` or run `npm test` yet. I'll update the PR description when I have.

File: shell/browser/api/electron_api_menu.cc [hunk 1]
```diff
@@ -281,10 +281,12 @@ std::u16string Menu::GetAcceleratorTextAtForTesting(int index) const {
 }
 
 void Menu::OnMenuWillClose() {
+  keep_alive_.Clear();
   Emit("menu-will-close");
```
Comment (deepak1556) [inline shell/browser/api/electron_api_menu.cc]:
```suggestion
 Emit("menu-will-close");
 keep_alive_.Clear();
```

**Review by deepak1556 — COMMENTED** [review]
Thanks for the detailed report! LGTM with a minor nit

A test would require surfacing a test only api similar to `SimulateSubmenuCloseSequenceForTesting` to open the application menu while gc happens, it can be a followup. I am fine merging in the current state.

**Comment by rf-figma** [comment]
Oh, thanks for fixing the tags! I think we do need to target 41 though, since it's still supported and the bug was backported to it.

**Review by deepak1556 — APPROVED** [review]
LGTM

**Review by ckerr — APPROVED** [review]
